modules/navier_stokes/src/fvkernels/INSFVTurbulentAdvection.C

```diff
@@ -0,0 +1,147 @@
+
+InputParameters
+INSFVTurbulentAdvection::validParams()
+{
+  auto params = INSFVAdvectionKernel::validParams();
+  params.addClassDescription(
+      "Advects an arbitrary turbulent quantity, the associated nonlinear 'variable'.");
+  params.addRequiredParam<MooseFunctorName>(NS::density, "fluid density");
+  params.addParam<std::vector<BoundaryName>>(
+      "walls", {}, "Boundaries that correspond to solid walls.");
+  return params;
+}
+
+INSFVTurbulentAdvection::INSFVTurbulentAdvection(const InputParameters & params)
+  : INSFVAdvectionKernel(params),
+    _rho(getFunctor<ADReal>(NS::density)),
+    _wall_boundary_names(getParam<std::vector<BoundaryName>>("walls"))
+{
+}
+
+void
+INSFVTurbulentAdvection::initialSetup()
+{
+  INSFVAdvectionKernel::initialSetup();
+  _wall_bounded = NS::getWallBoundedElements(_wall_boundary_names, _fe_problem, _subproblem);
+}
+
+ADReal
+INSFVTurbulentAdvection::computeQpResidual()
+{
+  const auto v = _rc_vel_provider.getVelocity(
+      _velocity_interp_method, *_face_info, determineState(), _tid, false);
+  const auto var_face = _var(makeFace(*_face_info,
+                                      limiterType(_advected_interp_method),
+                                      MetaPhysicL::raw_value(v) * _normal > 0),
+                             determineState());
+  const auto rho_face = _rho(makeFace(*_face_info,
+                                      limiterType(_advected_interp_method),
+                                      MetaPhysicL::raw_value(v) * _normal > 0),
+                             determineState());
+  return _normal * MetaPhysicL::raw_value(v) * rho_face * var_face;
+}
+
+void
+INSFVTurbulentAdvection::computeResidual(const FaceInfo & fi)
+{
+
+  if (skipForBoundary(fi))
+    return;
+
+  _face_info = &fi;
+  _normal = fi.normal();
+  _face_type = _face_info->faceType(std::make_pair(_var.number(), _var.sys().number()));
+  auto r = MetaPhysicL::raw_value(fi.faceArea() * fi.faceCoord() * computeQpResidual());
+
+  const Elem * elem = fi.elemPtr();
+  const Elem * neighbor = fi.neighborPtr();
+  auto bounded_elem = _wall_bounded[elem];
+  auto bounded_neigh = _wall_bounded[neighbor];
+
+  if (_face_type == FaceInfo::VarFaceNeighbors::ELEM ||
+      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
+  {
+    if (!bounded_elem)
+    {
+      // residual contribution of this kernel to the elem element
+      prepareVectorTag(_assembly, _var.number());
+      _local_re(0) = r;
+      accumulateTaggedLocalResidual();
+    }
+  }
+  if (_face_type == FaceInfo::VarFaceNeighbors::NEIGHBOR ||
+      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
+  {
+    if (!bounded_neigh)
+    {
+      // residual contribution of this kernel to the neighbor element
+      prepareVectorTagNeighbor(_assembly, _var.number());
+      _local_re(0) = -r;
+      accumulateTaggedLocalResidual();
+    }
+  }
+}
+
+void
+INSFVTurbulentAdvection::computeJacobian(const FaceInfo & fi)
```

Comment:
Sad that we need to override these functions, but I am supportive of this for now.